File: src/main/java/com/github/svegon/capi/mixininterface/IStatusEffectFogModifier.java
package com.github.svegon.capi.mixininterface;

import com.github.svegon.capi.util.FogContext;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectInstance;

public interface IStatusEffectFogModifier {
    StatusEffect getStatusEffect();

    void applyStartEndModifier(FogContext fogData, LivingEntity entity,
                               StatusEffectInstance effect, float viewDistance, float tickDelta);

    default boolean shouldApply(LivingEntity entity, float tickDelta) {
        return entity.hasStatusEffect(this.getStatusEffect());
    }

    default float applyColorModifier(LivingEntity entity, StatusEffectInstance effect, float horizonShadingRatio,
                                     float tickDelta) {
        StatusEffectInstance statusEffectInstance = entity.getStatusEffect(this.getStatusEffect());

        if (statusEffectInstance != null) {
            horizonShadingRatio = statusEffectInstance.getDuration() < 20
                    ? 1.0f - (float)statusEffectInstance.getDuration() / 20.0f : 0.0f;
        }

        return horizonShadingRatio;
    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/mixininterface/IStatusEffectFogModifier.kt
package io.github.svegon.capi.mixininterface

import io.github.svegon.capi.util.FogContext
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.effect.StatusEffect
import net.minecraft.entity.effect.StatusEffectInstance

interface IStatusEffectFogModifier {
    @JvmField
    val statusEffect: StatusEffect?

    fun applyStartEndModifier(
        fogData: FogContext?, entity: LivingEntity?,
        effect: StatusEffectInstance?, viewDistance: Float, tickDelta: Float
    )

    fun shouldApply(entity: LivingEntity, tickDelta: Float): Boolean {
        return entity.hasStatusEffect(this.statusEffect)
    }

    fun applyColorModifier(
        entity: LivingEntity, effect: StatusEffectInstance?, horizonShadingRatio: Float,
        tickDelta: Float
    ): Float {
        var horizonShadingRatio = horizonShadingRatio
        val statusEffectInstance = entity.getStatusEffect(this.statusEffect)

        if (statusEffectInstance != null) {
            horizonShadingRatio = if (statusEffectInstance.duration < 20
            ) 1.0f - statusEffectInstance.duration.toFloat() / 20.0f else 0.0f
        }

        return horizonShadingRatio
    }
}
